Fix #22232: rake test tasks exit status code

arekf opened the pull request:
Rake test task sets exit status code, so the behavior
is matching rails test command.

eileencodes commented:
Hi @arekf thanks for the PR. Can you explain better in your commit message and the changelog what fix means. Ex: explain what was the broken behavior and what the new behavior achieves. I see the linked issue but people reading the changelog/commit later may not be able to access it to find out what the reason for this change is. Thanks!

arekf commented:
Hello @eileencodes,
thank you for reviewing my changes. I've amended my commit, hope this makes more sense now. Let me know if I can make it any better.

Thanks,
Arek

sgrif commented:
@arekf Can you rebase?

arekf commented:
@sgrif Done!

sgrif commented on railties/CHANGELOG.md:
```diff
@@ -20,6 +20,13 @@
 
     *Puneet Agarwal* and *Xavier Noria*
 
+*   Fix setting exit status code for rake test tasks. The exit status code
```
This should be at the top of the file.

arekf replied:
Sorry, give me a moment :)

arekf commented:
@sgrif There we go!

sgrif commented:
Just glancing over these changes, doesn't this cause issues if you try to run more than one task in a command? e.g. `rake test:units test:functionals`. Should we only `exit` if the result is non-zero?

arthurnn commented:
Yeah, as far as I remember we didnt add those exit exactly because of what @sgrif said.
cc @senny @kaspth

senny commented:
I think the `Rake::TestTask` we were using previous to the new runner also exited with a non-zero status. See https://github.com/ruby/rake/blob/master/lib/rake/testtask.rb#L109

@sgrif is right tough. We should not exit for `0` status codes. Otherwise chaining tasks won't work anymore.

arekf commented:
@sgrif @arthurnn @senny Sorry, you're right it would've broken chaining tasks.
Please let me know if I can amend this to `exit` only when status code is non-zero (and break tasks chain). Should there be any additional message provided when it's exiting?

senny commented:
@arekf please feel free to amend.

arekf commented:
Can someone have a look at my amendments?

kaspth commented on railties/lib/rails/test_unit/testing.rake:
```diff
@@ -7,7 +7,8 @@ task default: :test
 task :test do
   $: << "test"
-  Minitest.rake_run(["test"])
+  passed = Minitest.rake_run(["test"])
```
Let's just extend `rake_run`, it was made to encapsulate a rake run :smile:

arekf replied:
@kaspth Here you go :)

arekf committed:
Fix #22232: rake test tasks exit status code

The exit status code was not set when tests were fired with `rake`.
Now, it is being set and it matches behavior of running tests via `rails`
command (`rails test`), so no matter if `rake test` or `rails test` command
is used the exit code will be set.

kaspth commented:
Thanks!
